### ISSUE TITLE
[r] Bug in nested-close logic

### ISSUE THREAD

**Issue by johnkerl** [single-cell-data/TileDB-SOMA#1477]
**Describe the bug**
Closing an experiment (or any collection) should recursively close all of its open descendants. Yet there is a scenario below in which something doesn't quite connect.

**To Reproduce**

* Step 1: create a `./pbmc3k` SOMA experiment:
  * https://gist.github.com/johnkerl/5f2e26d17dc930053cebbcb9970ecbcf
* Step 2 (optional): do some queries on it (not necessary to repro the bug):
  * https://gist.github.com/johnkerl/bdcbd51e58a13b2daf788a43c04aeac5
* Step 3: add new `obsm` and write it back:
  * https://gist.github.com/johnkerl/0d1f29c643a19afc25a46ddda8041d3e
  * The extra `ms_rna_w$close()` [here](https://gist.github.com/johnkerl/0d1f29c643a19afc25a46ddda8041d3e#file-gistfile1-txt-L48) shouldn't be necessary, but without it, `obsm` doesn't appear

Note that group writes (including member-adds) don't take effect until the group is closed.

Also note that spdl-logging shows when the close happens, and if the `ms_rna_w$close()` is commented out, that spdl-logging shows that `pbmc3k/ms/RNA` isn't getting closed.

Hypothesis: the `private$.member_cache` isn't including everything it should at https://github.com/single-cell-data/TileDB-SOMA/blob/21520a6c0ff922d272d25afa692cc48e2ea085f6/apis/r/R/TileDBGroup.R#L90

cc @aaronwolen @mojaveazure @mlin @pablo-gar 

Also note this closing logic is my own from https://github.com/single-cell-data/TileDB-SOMA/issues/1346 -- so I may be the best person to fix this. Or, maybe a fresh pair of eyes is best ...

**Versions (please complete the following information):**
 - TileDB-SOMA version: today's `main`
 - Language and language version: R
 - OS: reproduced on MacOS and Linux both

**Comment by eddelbuettel** [comment]
I see the assignment but I am really not that deep into R6 magic.  Help would be appreciate.  Maybe @mojaveazure and I can poke together...

**Comment by johnkerl** [comment]
@mlin in the spirit of load-balancing can you help take a look here?

**Comment by johnkerl** [comment]
(Also taking a peek, as I'd promised @eddelbuettel)

Following the repro in this issue's description, and using the third gist, and with this diff applied
https://gist.github.com/johnkerl/556f30a0c9d36e1c95ee01c34efc7a72
I see that what I _feared_ was true, _is_ true.

Namely when I print `exp$memfoo()` (dumper/inspector from the gist above) I see fewer things than when I do `ms_rna_w$memfoo()`.

This in turn suggests that perhaps that as called from that third gist's

```
exp <- SOMAExperimentOpen(exp$uri, "WRITE")
ms_rna_w <- exp$ms$get("RNA")
```

the `ms_rna_w` _appears_ to not be a reference to the `exp$ms$get("RNA")`.

I am made aware that S6 does have reference semantics, but, perhaps at some point in our implementation inheritance chain for the `get` method, we've inadvertently copying something -- ?

**Comment by eddelbuettel** [comment]
I actually spent several hours noodling around this issue, and have no fix.  Adding simple `on.exit(soma_parent$close())` closes too soon for our nesting so I have no suggested fix.

**Comment by johnkerl** [comment]
A bit more background:

* First gist in this PR's descriptions writes the SOMA experiment `./pbmc3k`
* The full structure looks like this
```
root group pbmc3k
  obs array
  ms group
    RNA group
      var array
      X group
        counts array
```
* The third gist does
```
...
pbmc3k <- RunPCA(pbmc3k, features = VariableFeatures(pbmc3k), verbose = FALSE)
...
exp$close()
exp <- SOMAExperimentOpen(exp$uri, "WRITE")
ms_rna_w <- exp$ms$get("RNA")
...
write_soma(pbmc3k[["pca"]], soma_parent = ms_rna_w) # This updates the `obsm`
...
```
* Using print statements, we find that
  * `exp`'s member cache shows what is open for the `exp` R6 object is only `obs` and `ms` in its member cache.
  * `ms_rna_w`'s member cache shows what is open for the `ms_rna_w` R6 object includes the new `obsm` as well
* Crucially, TileDB groups have _commit-on-close semantics_
  * So if we only call `exp$close()` (implementation [here](https://github.com/single-cell-data/TileDB-SOMA/blob/ee266fbeec9b84178ca74be802dd6f0774ed10f4/apis/r/R/TileDBGroup.R#L85-L106) then the new `obsm` is not persisted to storage, since the `exp` R6 object is not aware of it
  * If we do call `ms_rna_w$close()` then the new `obsm` _is_ persisted to storage, since the `ms_rna_w` R6 objecy is aware of it
  * In this latter case we successfully see
```
root group pbmc3k
  obs array
  ms group
    RNA group
      var array
      X group
        counts array
      obsm group    # <----- new
        X_pca array # <----- new
```
* Since the `ms_rna_w` R6 object was gotten from the `exp` R6 object via `ms_rna_w <- exp$ms$get("RNA")`, and since (naively) I believe R6 objects have reference semantics, the `exp` R6 object's member cache ought to traverse on `exp$close()` down to the same location as `ms_rna_w`, but, we see evidently that it does not

So there is an opportunity to do more debugging about R6 reference semantics.

**Comment by eddelbuettel** [comment]
Now that's a nice write-up!  (I used `spdl::*` rather than `print` myself but details).  One quick idea while reading this: "TileDB groups have commit-on-close semantics".  Could we work around this here (where we know things are nested) by _explicitly_ calling a commit or write rather than waiting for the _implicit_ one that happens too late?

**Comment by johnkerl** [comment]
> by explicitly calling a commit or write rather than waiting for the implicit one that happens too late?

@eddelbuettel I love this too, but unfortunately, the "commit" or "write" function for tiledb groups is, in fact, spelled c-l-o-s-e. So our only protection here would be to call `group$close()` before the user requested it (and without them having requested it). Which, I think, is not a good user experience ...

**Comment by eddelbuettel** [comment]
Maybe we have to add a french variant `e-c-r-i-r-e` then.

Kidding aside maybe the cost of close-then-reopen is what we have to bear here.  Better than failing.

**Comment by johnkerl** [comment]
Found and fixed in sandbox. PR coming up tomorrow.

**Comment by eddelbuettel** [comment]
What was the issue?  In a real-quick reader's digest version?

**Comment by johnkerl** [comment]
This simple repro shows the bug. First make a `getfoo` and `setfoo` at the `TileDBObject` level:

```diff
diff --git a/apis/r/R/TileDBObject.R b/apis/r/R/TileDBObject.R
index 7907d385..cc53d9ae 100644
--- a/apis/r/R/TileDBObject.R
+++ b/apis/r/R/TileDBObject.R
@@ -46,6 +46,13 @@ TileDBObject <- R6::R6Class(
       spdl::debug("[TileDBObject] initialize {} with '{}'", self$class(), self$uri)
     },
 
+    getfoo = function() {
+      return(private$.foo)
+    },
+    setfoo = function(value) {
+      private$.foo <- value
+    },
+
     #' @description Print the name of the R6 class.
     class = function() {
       class(self)[1]
@@ -121,6 +128,8 @@ TileDBObject <- R6::R6Class(
   ),
 
   private = list(
+    .foo = NULL,
+
     # Pro tip: in R6 we can't set these to anything other than NULL here, even if we want to.  If
     # you want them defaulted to anything other than NULL, leave them NULL here and set the defaults
     # in the constructor.
```

Next try it out -- if `exp$ms` is the desired reference then we can set/get the `.foo` attribute on `exp` as well as `exp$ms`:

```R
> exp$setfoo(100)
> exp$getfoo()
[1] 100
# Works
```

```R
> exp$ms$setfoo(200)
> exp$ms$getfoo()
NULL
# Not good
```

```R
> ms <- exp$ms
> ms$getfoo()
NULL
> ms$setfoo(300)
> ms$getfoo()
[1] 300
# Readback works as long as we retain `ms <- exp$ms` 
```

The bug is simple: on `TileDBGroup`'s `set` we do update the member cache here
https://github.com/single-cell-data/TileDB-SOMA/blob/ee266fbeec9b84178ca74be802dd6f0774ed10f4/apis/r/R/TileDBGroup.R#L139-L144
whereas on `TileDBGroup`'s `get` we do not:
https://github.com/single-cell-data/TileDB-SOMA/blob/ee266fbeec9b84178ca74be802dd6f0774ed10f4/apis/r/R/TileDBGroup.R#L173-L183

The fix is simple:
```
private$add_cached_member(name, obj)
```

in `TileDBGroup`'s `get`

**Comment by eddelbuettel** [comment]
For what it is worth I did create a local script reproducing this issue (and hinging on the undesirable explicit `ms_rna_w$close()` to flip between 'works-with-the-extra-close' and 'does not work'.  Applying the fix suggested in #1524 improves this and we no longer need `ms_rna_w$close()` --- which is quite promising.

**Comment by johnkerl** [comment]
#1524 is merged
